File: packages/client/src/types/domains.ts
/**
 * Domain-specific types for all 13 Open Harness API domains
 */

import type {
  HarnessId,
  AgentId,
  SkillId,
  SessionId,
  ExecutionId,
  ISO8601,
  ExecutionStatus,
  FileInfo,
} from "./common";

// ============================================================================
// 1. Harness Registry
// ============================================================================

export interface Harness {
  id: HarnessId;
  name: string;
  version: string;
  vendor: string;
  status: HarnessStatus;
  capabilities: string[];
  created_at: ISO8601;
}

export type HarnessStatus =
  | "active"
  | "inactive"
  | "maintenance"
  | "deprecated";

export interface HarnessCredentials {
  api_key?: string;
  oauth?: {
    client_id: string;
    client_secret: string;
  };
}

export interface CapabilityManifest {
  harness_id: HarnessId;
  version: string;
  capabilities: CapabilityDeclaration[];
}

export interface CapabilityDeclaration {
  id: string;
  supported: boolean;
  constraints?: Record<string, unknown>;
}

// ============================================================================
// 2. Agents (OAF-compliant)
// ============================================================================

export interface ModelConfig {
  provider: string;
  name: string;
  embedding?: string;
}

export interface Agent {
  id: AgentId;
  // OAF identity fields
  name: string;
  vendorKey: string;
  agentKey: string;
  version: string;
  slug: string;
  // Metadata
  description: string;
  author?: string;
  license?: string;
  tags: string[];
  // Configuration
  model: string | ModelConfig;
  system_prompt?: string;
  skills: SkillId[];
  created_at: ISO8601;
  updated_at: ISO8601;
}

export interface CreateAgentRequest {
  name: string;
  vendorKey?: string;
  agentKey?: string;
  version?: string;
  slug?: string;
  description?: string;
  author?: string;
  license?: string;
  tags?: string[];
  model: string | ModelConfig;
  system_prompt?: string;
  skills?: SkillId[];
}

export interface UpdateAgentRequest {
  name?: string;
  vendorKey?: string;
  agentKey?: string;
  version?: string;
  slug?: string;
  description?: string;
  author?: string;
  license?: string;
  tags?: string[];
  model?: string | ModelConfig;
  system_prompt?: string;
  skills?: SkillId[];
}

// ============================================================================
// Agent Import/Export (OAF-compliant)
// ============================================================================

export type PackageContentsMode = "bundled" | "referenced";

export interface ExportAgentRequest {
  include_memory?: boolean;
  include_versions?: boolean;
  contents_mode?: PackageContentsMode;
}

export interface ImportAgentRequest {
  rename_to?: string;
  merge_strategy?: "fail" | "overwrite" | "skip";
}

export interface ImportAgentResponse {
  agent: Agent;
  warnings: string[];
}

// ============================================================================
// 3. Skills
// ============================================================================

export interface Skill {
  id: SkillId;
  name: string;
  version: string;
  description: string;
  author?: string;
  tools: SkillTool[];
  installed_at: ISO8601;
}

export interface SkillTool {
  name: string;
  description: string;
  input_schema: object;
}

export interface RegisterSkillRequest {
  source: SkillSource;
  config?: Record<string, unknown>;
}

export type SkillSource =
  | { type: "url"; url: string }
  | { type: "file"; path: string }
  | { type: "inline"; content: string };

// ============================================================================
// 4. MCP Servers
// ============================================================================

export interface McpServer {
  id: string;
  name: string;
  status: "connected" | "disconnected" | "error";
  transport: McpTransport;
  tools_count: number;
  resources_count: number;
  prompts_count: number;
  connected_at?: ISO8601;
}

export type McpTransport =
  | { type: "stdio"; command: string; args?: string[] }
  | { type: "http"; url: string }
  | { type: "sse"; url: string };

export interface ConnectMcpServerRequest {
  name: string;
  transport: McpTransport;
  auto_reconnect?: boolean;
}

export interface McpTool {
  name: string;
  description: string;
  input_schema: object;
}

export interface McpResource {
  uri: string;
  name: string;
  description?: string;
  mime_type?: string;
}

export interface McpPrompt {
  name: string;
  description?: string;
  arguments?: McpPromptArgument[];
}

export interface McpPromptArgument {
  name: string;
  description?: string;
  required?: boolean;
}

// ============================================================================
// 5. Tools
// ============================================================================

export interface Tool {
  id: string;
  name: string;
  description: string;
  source: "builtin" | "mcp" | "skill" | "custom";
  source_id?: string;
  input_schema: object;
}

export interface RegisterToolRequest {
  name: string;
  description: string;
  input_schema: object;
  handler: ToolHandler;
}

export type ToolHandler =
  | { type: "webhook"; url: string }
  | { type: "command"; command: string; args?: string[] };

export interface InvokeToolResponse {
  success: boolean;
  output: object;
  duration_ms: number;
  error?: string;
}

// ============================================================================
// 6. Execution
// ============================================================================

export interface Execution {
  id: ExecutionId;
  harness_id: HarnessId;
  agent_id?: AgentId;
  status: ExecutionStatus;
  message: string;
  started_at: ISO8601;
  completed_at?: ISO8601;
}

export interface ExecuteRequest {
  message: string;
  agent_id?: AgentId;
  skills?: SkillId[];
  model?: string;
  max_tokens?: number;
  temperature?: number;
  system_prompt?: string;
  session_id?: SessionId;
}

export interface ExecuteResponse {
  execution_id: ExecutionId;
  status: ExecutionStatus;
  stream_url: string;
}

// ============================================================================
// 7. Sessions
// ============================================================================

export interface Session {
  id: SessionId;
  name?: string;
  status: "active" | "paused" | "ended";
  agent_id?: AgentId;
  message_count: number;
  created_at: ISO8601;
  updated_at: ISO8601;
}

export interface CreateSessionRequest {
  name?: string;
  agent_id?: AgentId;
  skills?: SkillId[];
  system_prompt?: string;
}

export interface CreateSessionResponse {
  session: Session;
  connect_url: string;
}

// ============================================================================
// 8. Memory
// ============================================================================

export interface MemoryState {
  agent_id: AgentId;
  blocks: MemoryBlock[];
  archive_size: number;
}

export interface MemoryBlock {
  label: string;
  value: string;
  read_only: boolean;
  updated_at: ISO8601;
}

export interface CreateMemoryBlockRequest {
  label: string;
  value: string;
  read_only?: boolean;
}

export interface MemorySearchResult {
  source: "block" | "archive";
  label?: string;
  content: string;
  relevance_score: number;
}

export interface ArchiveEntry {
  id: string;
  content: string;
  created_at: ISO8601;
  metadata?: Record<string, unknown>;
}

// ============================================================================
// Memory Import/Export
// ============================================================================

export type MemoryMergeStrategy = "overwrite" | "skip" | "merge";

export interface ExportMemoryRequest {
  include_archive?: boolean;
}

export interface ImportMemoryRequest {
  merge_strategy?: MemoryMergeStrategy;
}

export interface ImportMemoryResponse {
  blocks_imported: number;
  archive_entries_imported: number;
  conflicts: number;
  warnings: string[];
}

// ============================================================================
// 9. Subagents
// ============================================================================

export interface Subagent {
  id: string;
  name: string;
  description: string;
  status: "idle" | "running" | "completed" | "failed";
  parent_agent_id: AgentId;
  created_at: ISO8601;
}

export interface SpawnSubagentRequest {
  name: string;
  description: string;
  system_prompt?: string;
  skills?: SkillId[];
  model?: string;
}

export interface DelegateTaskRequest {
  task: string;
  context?: Record<string, unknown>;
}

// ============================================================================
// 10. Files
// ============================================================================

export interface SearchFilesRequest {
  glob?: string;
  grep?: string;
  path?: string;
  max_results?: number;
}

export interface FileMatch {
  file: FileInfo;
  line_matches?: LineMatch[];
}

export interface LineMatch {
  line_number: number;
  content: string;
}

export interface UploadFileRequest {
  path: string;
  file: Blob | Buffer;
  overwrite?: boolean;
}

// ============================================================================
// 11. Hooks & Events
// ============================================================================

export interface Hook {
  id: string;
  event: HookEvent;
  handler: HookHandler;
  enabled: boolean;
  created_at: ISO8601;
}

export type HookEvent =
  | "execution.start"
  | "execution.end"
  | "tool.before"
  | "tool.after"
  | "session.start"
  | "session.end"
  | "error";

export type HookHandler =
  | { type: "webhook"; url: string }
  | { type: "command"; command: string; args?: string[] };

export interface RegisterHookRequest {
  event: HookEvent;
  handler: HookHandler;
  enabled?: boolean;
}

export interface Webhook {
  id: string;
  url: string;
  events: string[];
  secret?: string;
  created_at: ISO8601;
}

// ============================================================================
// 12. Planning
// ============================================================================

export interface Plan {
  execution_id: ExecutionId;
  tasks: import("./events").PlanTask[];
  updated_at: ISO8601;
}

// ============================================================================
// 13. Conformance & Diagnostics
// ============================================================================

export interface ConformanceRun {
  id: string;
  harness_version: string;
  result: "pass" | "fail" | "partial";
  passed: number;
  failed: number;
  skipped: number;
  golden_rule_violations: number;
  run_at: ISO8601;
}

export interface ConformanceStatus {
  status: "certified" | "partial" | "failing" | "not_tested";
  certified_version?: string;
  pass_rate: number;
  last_run_at?: ISO8601;
  golden_rule_compliant: boolean;
}

export interface DiagnosticsResponse {
  harness_id: HarnessId;
  version: string;
  uptime_seconds: number;
  memory_usage_mb: number;
  active_sessions: number;
  active_executions: number;
  connected_mcp_servers: number;
  installed_skills: number;
  config: Record<string, unknown>;
}

export interface LogEntry {
  timestamp: ISO8601;
  level: "debug" | "info" | "warn" | "error";
  message: string;
  context?: Record<string, unknown>;
}

File: packages/python/src/openharness/types/domains.py
"""
Domain-specific types for the Open Harness API.
"""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field

from .common import (
    AgentId,
    ExecutionId,
    HookId,
    McpServerId,
    MemoryBlockId,
    SessionId,
    SkillId,
    SubagentId,
    TodoId,
    Tool,
    UsageStats,
)


# =============================================================================
# Agents Domain
# =============================================================================


class AgentState(str, Enum):
    """Agent state."""
    ACTIVE = "active"
    PAUSED = "paused"
    ARCHIVED = "archived"


class ModelConfig(BaseModel):
    """Model configuration following OAF spec."""
    provider: str | None = None
    name: str | None = None
    embedding: str | None = None


class Agent(BaseModel):
    """Agent definition following Open Agent Format (OAF) spec."""
    id: AgentId
    name: str
    # OAF identity fields
    vendor_key: str | None = Field(default=None, alias="vendorKey")
    agent_key: str | None = Field(default=None, alias="agentKey")
    version: str | None = None
    slug: str | None = None
    # Metadata
    description: str | None = None
    author: str | None = None
    license: str | None = None
    tags: list[str] = Field(default_factory=list)
    # Configuration
    system_prompt: str | None = None
    model: str | ModelConfig | None = None
    state: AgentState = AgentState.ACTIVE
    tools: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime

    class Config:
        populate_by_name = True


class CreateAgentRequest(BaseModel):
    """Request to create an agent."""
    name: str
    # OAF identity fields
    vendor_key: str | None = Field(default=None, alias="vendorKey")
    agent_key: str | None = Field(default=None, alias="agentKey")
    version: str | None = None
    slug: str | None = None
    # Metadata
    description: str | None = None
    author: str | None = None
    license: str | None = None
    tags: list[str] = Field(default_factory=list)
    # Configuration
    system_prompt: str | None = None
    model: str | ModelConfig | None = None
    tools: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)

    class Config:
        populate_by_name = True


class UpdateAgentRequest(BaseModel):
    """Request to update an agent."""
    name: str | None = None
    vendor_key: str | None = Field(default=None, alias="vendorKey")
    agent_key: str | None = Field(default=None, alias="agentKey")
    version: str | None = None
    slug: str | None = None
    description: str | None = None
    author: str | None = None
    license: str | None = None
    tags: list[str] | None = None
    system_prompt: str | None = None
    model: str | ModelConfig | None = None
    state: AgentState | None = None
    tools: list[str] | None = None
    metadata: dict[str, Any] | None = None

    class Config:
        populate_by_name = True


# =============================================================================
# Agent Import/Export (OAF-compliant)
# =============================================================================


class PackageContentsMode(str, Enum):
    """Package contents mode for OAF ZIP files."""
    BUNDLED = "bundled"      # All skills included, works offline
    REFERENCED = "referenced"  # Skills fetched from URLs at install time


class ExportAgentRequest(BaseModel):
    """Request to export an agent as OAF package."""
    include_memory: bool = False
    include_versions: bool = False
    contents_mode: PackageContentsMode = PackageContentsMode.BUNDLED


class ImportAgentRequest(BaseModel):
    """Request to import an agent from OAF package."""
    rename_to: str | None = None
    merge_strategy: str = "fail"  # "fail", "overwrite", "skip"


class ImportAgentResponse(BaseModel):
    """Response from agent import."""
    agent: "Agent"
    warnings: list[str] = Field(default_factory=list)


# =============================================================================
# Memory Import/Export
# =============================================================================


class MemoryMergeStrategy(str, Enum):
    """Memory import merge strategy."""
    OVERWRITE = "overwrite"
    SKIP = "skip"
    MERGE = "merge"


class ExportMemoryRequest(BaseModel):
    """Request to export agent memory."""
    include_archive: bool = True


class ImportMemoryRequest(BaseModel):
    """Request to import agent memory."""
    merge_strategy: MemoryMergeStrategy = MemoryMergeStrategy.OVERWRITE


class ImportMemoryResponse(BaseModel):
    """Response from memory import."""
    blocks_imported: int
    archive_entries_imported: int
    conflicts: int = 0
    warnings: list[str] = Field(default_factory=list)


# =============================================================================
# Skills Domain
# =============================================================================


class SkillStatus(str, Enum):
    """Skill status."""
    ACTIVE = "active"
    DISABLED = "disabled"
    ERROR = "error"


class Skill(BaseModel):
    """Skill definition."""
    id: SkillId
    name: str
    description: str
    version: str
    status: SkillStatus = SkillStatus.ACTIVE
    source: str  # "registry", "local", "url"
    tools: list[Tool] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    installed_at: datetime


class InstallSkillRequest(BaseModel):
    """Request to install a skill."""
    source: str
    version: str | None = None
    config: dict[str, Any] = Field(default_factory=dict)


# =============================================================================
# MCP Domain
# =============================================================================


class McpServerStatus(str, Enum):
    """MCP server connection status."""
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    ERROR = "error"
    CONNECTING = "connecting"


class McpServer(BaseModel):
    """MCP server connection."""
    id: McpServerId
    name: str
    command: str
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)
    status: McpServerStatus = McpServerStatus.DISCONNECTED
    tools: list[Tool] = Field(default_factory=list)
    connected_at: datetime | None = None


class ConnectMcpRequest(BaseModel):
    """Request to connect to an MCP server."""
    name: str
    command: str
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)


class McpResource(BaseModel):
    """MCP resource."""
    uri: str
    name: str
    description: str | None = None
    mime_type: str | None = None


class McpPrompt(BaseModel):
    """MCP prompt template."""
    name: str
    description: str | None = None
    arguments: list[dict[str, Any]] = Field(default_factory=list)


# =============================================================================
# Execution Domain
# =============================================================================


class ExecutionStatus(str, Enum):
    """Execution status."""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class ExecuteRequest(BaseModel):
    """Request to execute a prompt."""
    message: str
    agent_id: AgentId | None = None
    session_id: SessionId | None = None
    system_prompt: str | None = None
    tools: list[str] | None = None
    model: str | None = None
    max_tokens: int | None = None
    temperature: float | None = Field(default=None, ge=0, le=1)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Execution(BaseModel):
    """Execution record."""
    id: ExecutionId
    status: ExecutionStatus
    message: str
    response: str | None = None
    agent_id: AgentId | None = None
    session_id: SessionId | None = None
    usage: UsageStats | None = None
    tool_calls: list["ToolCall"] = Field(default_factory=list)
    created_at: datetime
    completed_at: datetime | None = None


class ToolCall(BaseModel):
    """Tool call record."""
    id: str
    name: str
    input: dict[str, Any]
    output: Any | None = None
    error: str | None = None
    duration_ms: int | None = None


# =============================================================================
# Sessions Domain
# =============================================================================


class SessionStatus(str, Enum):
    """Session status."""
    ACTIVE = "active"
    ARCHIVED = "archived"


class Session(BaseModel):
    """Conversation session."""
    id: SessionId
    name: str | None = None
    agent_id: AgentId | None = None
    status: SessionStatus = SessionStatus.ACTIVE
    system_prompt: str | None = None
    message_count: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class CreateSessionRequest(BaseModel):
    """Request to create a session."""
    name: str | None = None
    agent_id: AgentId | None = None
    system_prompt: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class Message(BaseModel):
    """Message in a session."""
    id: str
    role: str  # "user", "assistant", "system"
    content: str
    tool_calls: list[ToolCall] = Field(default_factory=list)
    created_at: datetime


# =============================================================================
# Memory Domain
# =============================================================================


class MemoryBlockType(str, Enum):
    """Memory block type."""
    CORE = "core"
    ARCHIVAL = "archival"
    WORKING = "working"
    CUSTOM = "custom"


class MemoryBlock(BaseModel):
    """Memory block."""
    id: MemoryBlockId
    type: MemoryBlockType
    label: str
    content: str
    read_only: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class CreateMemoryBlockRequest(BaseModel):
    """Request to create a memory block."""
    type: MemoryBlockType
    label: str
    content: str
    read_only: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)


class UpdateMemoryBlockRequest(BaseModel):
    """Request to update a memory block."""
    content: str | None = None
    label: str | None = None
    metadata: dict[str, Any] | None = None


class ArchiveEntry(BaseModel):
    """Archive memory entry."""
    id: str
    content: str
    embedding: list[float] | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


# =============================================================================
# Subagents Domain
# =============================================================================


class SubagentStatus(str, Enum):
    """Subagent status."""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class Subagent(BaseModel):
    """Spawned subagent."""
    id: SubagentId
    type: str
    prompt: str
    status: SubagentStatus = SubagentStatus.PENDING
    result: Any | None = None
    error: str | None = None
    parent_id: SubagentId | None = None
    created_at: datetime
    completed_at: datetime | None = None


class SpawnSubagentRequest(BaseModel):
    """Request to spawn a subagent."""
    type: str
    prompt: str
    config: dict[str, Any] = Field(default_factory=dict)


# =============================================================================
# Files Domain
# =============================================================================


class FileInfo(BaseModel):
    """File information."""
    path: str
    name: str
    size: int
    is_directory: bool
    mime_type: str | None = None
    modified_at: datetime | None = None


class ReadFileRequest(BaseModel):
    """Request to read a file."""
    path: str
    offset: int | None = None
    limit: int | None = None


class WriteFileRequest(BaseModel):
    """Request to write a file."""
    path: str
    content: str | bytes
    create_directories: bool = False


class SearchFilesRequest(BaseModel):
    """Request to search files."""
    pattern: str
    path: str | None = None
    max_results: int = 100


# =============================================================================
# Hooks Domain
# =============================================================================


class HookType(str, Enum):
    """Hook type."""
    PRE_TOOL = "pre_tool"
    POST_TOOL = "post_tool"
    STOP = "stop"
    CUSTOM = "custom"


class Hook(BaseModel):
    """Hook definition."""
    id: HookId
    type: HookType
    name: str
    command: str
    enabled: bool = True
    config: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class CreateHookRequest(BaseModel):
    """Request to create a hook."""
    type: HookType
    name: str
    command: str
    config: dict[str, Any] = Field(default_factory=dict)


# =============================================================================
# Planning Domain
# =============================================================================


class TodoStatus(str, Enum):
    """Todo status."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"


class Todo(BaseModel):
    """Todo item."""
    id: TodoId
    content: str
    status: TodoStatus = TodoStatus.PENDING
    priority: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class CreateTodoRequest(BaseModel):
    """Request to create a todo."""
    content: str
    priority: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)


class UpdateTodoRequest(BaseModel):
    """Request to update a todo."""
    content: str | None = None
    status: TodoStatus | None = None
    priority: int | None = None
    metadata: dict[str, Any] | None = None


# =============================================================================
# Models Domain
# =============================================================================


class ModelInfo(BaseModel):
    """Model information."""
    id: str
    name: str
    provider: str
    description: str | None = None
    context_window: int | None = None
    max_output_tokens: int | None = None
    supports_vision: bool = False
    supports_tools: bool = False
    supports_streaming: bool = True
